### PR TITLE
Driver-specific spinlock for cxd56_serial.c

### PR DESCRIPTION
**Pull request by masayuki2009**
## Summary

- This PR consists of the following 2 commits
- commit 1: include: nuttx: Introduce spinlock_t for #ifndef CONFIG_SPINLOCK
   - This commit introduces spinlock_t for #ifndef CONFIG_SPINLOCK
      which is useful for the non-SMP case because it does not consume
      memory
- commit 2: arch: cxd56xx: Introduce driver specific spinlock in cxd56_serial.c
  - This commit introduces driver-specific spinlock in cxd56_serial.c
      to improve performance
- NOTE: https://github.com/apache/incubator-nuttx/issues/2213

## Impact

- SMP only

## Testing

- Tested with spresense:wifi and spresense:wifi_smp

### CODE REVIEW

File: include/nuttx/spinlock.h [hunk 1]
```diff
@@ -30,6 +30,12 @@
 #include <sys/types.h>
 #include <stdint.h>
 
+#ifndef CONFIG_SPINLOCK
+typedef struct
+{
+} spinlock_t;
+#endif
+
 #ifdef CONFIG_SPINLOCK
```
Comment (xiaoxiang781216) [inline include/nuttx/spinlock.h:39]: change:
```
#endif

#ifdef CONFIG_SPINLOCK
```
to
```
#else
```

Comment (masayuki2009) [reply include/nuttx/spinlock.h]: @xiaoxiang781216 

Done

**Review by xiaoxiang781216 — APPROVED** [review]

File: include/nuttx/spinlock.h [hunk 2]
```diff
@@ -30,7 +30,11 @@
 #include <sys/types.h>
 #include <stdint.h>
 
-#ifdef CONFIG_SPINLOCK
+#ifndef CONFIG_SPINLOCK
```
Comment (gustavonihei) [inline include/nuttx/spinlock.h:33]: I'm trying to understand the outcome of this change when `CONFIG_SPINLOCK` is not defined.
From what I understood, `&priv->lock` should always evaluate to `NULL` when `CONFIG_SPINLOCK` is not defined.
I've tried to isolate this behavior, but I couldn't achieve the same result.

https://godbolt.org/z/467afd

Did I do something wrong in my test?

Comment (masayuki2009) [reply include/nuttx/spinlock.h:33]: Hi @gustavonihei,

Actually, when CONFIG_SPINLOCK is not defined CONFIG_SMP is not defined either.
So, finally, spin_lock_irqsave(l) will be just up_irq_save().

```
include/nuttx/irq.h

...
#if defined(CONFIG_SMP)                                                                                                                                                
irqstate_t spin_lock_irqsave(spinlock_t *lock);                                                                                                                        
#else                                                                                                                                                                  
#  define spin_lock_irqsave(l) up_irq_save()                                                                                                                           
#endif  
...
```

Comment (gustavonihei) [reply include/nuttx/spinlock.h:33]: Thanks for the clarification, @masayuki2009!

**Pull request merged**